### PR TITLE
fix(bzlmod pip.parse): allow requirements with duplicate package lines

### PR DESCRIPTION
**Pull request by rickeylev**
Requirements files are permitted to have duplicate lines for the same package. An example
of this is having separate lines for a package and its extras. When we parse requirements,
the parser discards the "extra" portion of the entry and returns a list of all the packages
as-is. When a repository is created for each entry, this means the same name is used for
multiple repositories.

Under WORKSPACE builds, duplicate repository names aren't an error. It appears that last
defined repo takes affect. Under bzlmod, duplicate repo names are an error.

To fix, mimic the last-defined-wins behavior in bzlmod by using a map to dedupe the package names.

Fixes https://github.com/bazelbuild/rules_python/issues/1615

### CODE REVIEW

File: python/private/bzlmod/pip.bzl [hunk 1]
```diff
@@ -109,7 +109,19 @@ def _create_whl_repos(module_ctx, pip_attr, whl_map, whl_overrides):
...
+    # Defining a repo with the same name multiple times, but only the last
+    # definition is respected.
+    # The requirement lines might have duplicate names because lines for extras
+    # are returned as just the base package name. e.g., `foo[bar]` results
+    # in an entry like `("foo", "foo[bar] == 1.0 ...")`.
+    requirements = {
+        normalize_name(entry[0]): entry
+        # The WORKSPACE pip_parse sorted entries, so mimic that ordering.
+        for entry in sorted(parse_result.requirements)
+    }.values()
```
Comment (aignas) [inline python/private/bzlmod/pip.bzl:123]: I was wondering if the fix should be in the `parse_requirements` function instead, but I do like it here better even though this needs an extra comment here.

**Review by aignas — APPROVED** [review]
Thanks for adding a test for this!

**Comment by aignas** [comment]
The test failures look like the `pre-commit` hook may need to be run or something similar?

**Comment by AndrewGuenther** [comment]
@rickeylev I patched this to our local environment and can confirm it works!

**Pull request merged**